enhancements/two-nodes-openshift/2no.md

```diff
@@ -0,0 +1,398 @@
+* Provide an OpenShift cluster experience that is identical to that of a 3-node hyperconverged cluster, but with 2 nodes
+
+### Non-Goals
+
+* Workload resilience - see related [Pre-DRAFT enhancement](https://docs.google.com/document/d/1TDU_4I4LP6Z9_HugeC-kaQ297YvqVJQhBs06lRIC9m8/edit)
+* Resilient storage - see future enhancement
+* Support for platforms other than bare metal including automated ci testing
+* Support for other topologies (eg. hypershift)
+* Adding worker nodes
+* Creation RHEL-HA events and metrics for consumption by the OpenShift monitoring stack (Deferred to post-MVP)
+
+## Proposal
+
+Use the RHEL-HA stack (Corosync, and Pacemaker), which has been used to delivered supported 2-node cluster experiences for multiple decades, to manage cri-o, kubelet, and the etcd daemon.
+etcd will run as as a voting member on both nodes.
+We will take advantage of RHEL-HA's native support for systemd and re-use the standard cri-o and kublet units, as well as create a new Open Cluster Framework (OCF) script for etcd.
+The existing startup order of cri-o, then kubelet, then etcd will be preserved.
+The `etcdctl`, `etcd-metrics`, and `etcd-readyz` containers will remain part of the static pod, the contents of which remains under the exclusive control of the Cluster Etcd Operator (CEO).
+
+Use RedFish compatible Baseboard Management Controllers (BMCs) as our primary mechanism to power off (fence) an unreachable peer and ensure that it can do no harm while the remaining node continues.
+
+Upon a peer failure, the RHEL-HA components on the surivor will fence the peer and use the OCF script to restart etcd as a new cluster-of-one.
+
+Upon a network failure, the RHEL-HA components ensure that exactly one node will survive, fence it's peer, and use the OCF script to restart etcd as a new cluster-of-one.
+
+In both cases, the control-plane's dependance on etcd will cause it to respond with errors until etcd has been restarted.
+
+Upon rebooting, the RHEL-HA components ensure that a node remains inert (not running cri-o, kubelet, or etcd) until it sees it's peer.
+If the failed peer is likely to remain offline for an extended period of time, admin confirmation is required on the remaining node to allow it to start OpenShift.
+The functionality exists within RHEL-HA, but a wrapper will be provided to take care of the details.
+
+When starting etcd, the OCF script will use etcd's cluster ID and version counter to determine whether the existing data directory can be reused, or must be erased before joining an active peer.
+
+### Workflow Description
+
+#### Cluster Creation
+
+User creation of a two node control-plane will be possible via the Assisted Installer.
+
+The procedure follows the standard flow except for the configuration of 2 nodes instead of 3, and the collection of RedFish details (including passwords!) for each node which are needed for the RHEL-HA configuration.
```

Comment:
Where does this `RedFish` information get collected? Will this be in the installconfig somewhere?
Is this something that should get called out in API changes?

Reply:
Machine CRs have a place for this already

Reply:
So are you saying this will only be added after installation is complete? I know that's something we discussed, but I wasn't sure a decision had been made.

Reply:
ABI today supports providing the redfish information in the platform:baremetal section of the install-config.
I believe it gets passed through assisted-service via the installConfigOverrides, since there is no API in assisted-service today to provide this info.
It ends up in the target cluster in the BMH CR (with credentials in a Secret linked from that CR).

If the plan is to use the BMH to store information about the hardware then I think this enhancement should say so explicitly. TBH, I don't think that actually works.
If the plan is to use the baremetal-operator to do the fencing then failover will be far too slow, because we will have to reschedule BMO on the live node before we can be sure of fencing the other node.
If the plan is _not_ to use the baremetal-operator, then we don't want to give BMO credentials for the hosts, because ironic will fight against anything else trying to use Redfish to change the power state.
(Maybe we have to power off _both_ in Redfish directly and via BMO?)

Either way, this should be called out.

Reply:
@mshitrit what's the direction here?

Reply:
At the moment I think we are leaning into applying the `RedFish` information by the new operator, but this decision isn't finalized yet